### PR TITLE
ENH: Performance improvements in scipy.linalg.special_matrices

### PR DESCRIPTION
**Pull request by droyed**
My first PR here. I care about performance, so hopefully my requests would be focused on that aspect. I figured SciPy could use some performance improvements.

Basically two folds of changes with this PR for four functions : 

I. scipy.linalg.special_matrices.py/tri : The crux of the implementation is just one line, which is improved with a better use of memory. The performance scales very well across all datasizes, with a speedup over `2.3x` on decent and higher sized arrays. Some timings are presented next for reference -

	In [1190]: N, M, k = 5000, 5000, 10

	In [1192]: %timeit tri(N, M, k)
	      ...: %timeit np.greater_equal.outer(np.arange(k,N+k), np.arange(M))
	      ...: 
	10 loops, best of 3: 47.9 ms per loop
	100 loops, best of 3: 19.6 ms per loop

	In [1193]: N, M, k = 50, 50, 2

	In [1194]: %timeit tri(N, M, k)
	      ...: %timeit np.greater_equal.outer(np.arange(k,N+k), np.arange(M))
	      ...: 
	100000 loops, best of 3: 7.41 µs per loop
	100000 loops, best of 3: 5.4 µs per loop

II. Next up, we are using `numpy.lib.stride_tricks.as_strided` for striding and thus achieving performance benefits. This works well for `toeplitz`, `circulant` and `hankel`. Some timings -

a. scipy.linalg.special_matrices.py/circulant : `4x+` speedup here and scales well again.

	In [40]: from scipy.linalg import circulant

	In [41]: a = np.random.randint(0,9,(5000))

	In [42]: %timeit circulant(a)
	10 loops, best of 3: 75.5 ms per loop

	In [43]: %timeit circulant_new(a)
	100 loops, best of 3: 18.8 ms per loop

	In [44]: a = np.random.randint(0,9,(50))

	In [45]: %timeit circulant(a)
	10000 loops, best of 3: 21.8 µs per loop

	In [46]: %timeit circulant_new(a)
	100000 loops, best of 3: 9.77 µs per loop

b. scipy.linalg.special_matrices.py/toeplitz : Similar numbers on performance.

c. scipy.linalg.special_matrices.py/hankel : Similar numbers on performance.

### CODE REVIEW

**Comment by droyed** [comment]
Seems like I missed out editing one comment in scipy.linalg.special_matrices.py at lines 198,199 for the function `toeplitz` and should be the following - 
    
    # Form a 1D array containing a reversed c followed by r[1:] that could be 
    # strided to give us toeplitz matrix

**Comment by ilayn** [comment]
I haven't tested thoroughly but if others are happy with playing with stride fire, it looks good to me. 

The SO reference is not needed though. It's just an explanation of the code but not necessarily the algorithm or the theory which is not applicable here anyways.

**Comment by droyed** [comment]
@ilayn Thanks for the feedback! Well, on the reference, since most of the code is taken from the referenced SO post, I though it's a good practice to provide the attribution. Seen [`few`](https://github.com/scipy/scipy/blob/ad4bf5d439668d2f2ce12c0ff56611777c80f880/scipy/signal/spectral.py#L1707-L1724) before.

**Comment by ilayn** [comment]
Ah let me reword it. I didn't make myself clear. Referencing is of course very important. 

Imagine a user clicks that link from the documentation and goes to your answer. S/he won't see the relation between circulant matrices and that SO post unless goes into the source code. That's what I meant. Otherwise SO or not online sources are also quite valuable. Your example is also inside a private function so it clarifies an implementation detail.

**Comment by pv** [comment]
It's also useful to note that everyone invents nice tricks that they 
reuse when writing code. Usually, these don't cross the line of being 
noteworthy enough for needing a citation.

**Comment by droyed** [comment]
@ilayn @pv I guess it doesn't matter in this case, because I am the SO poster as well. Should I edit out that `Reference` section?

**Comment by rgommers** [comment]
@droyed yes, please remove the SO reference

Also, given that this is a performance improvement, you may want to add a benchmark that covers the speedup. It would go in ``benchmarks/benchmarks/linalg.py``

**Comment by droyed** [comment]
@rgommers On the timing results that I showed earlier were done in a separate script in which I had both the current version of SciPy (without my commits) and my commits as two separate functions. For example, I had : `circulant` (SciPy current version without my changes) and `circulant_new` being my committed implementation. So, to replicate those same benchmarking steps, I would then need to keep both the versions. I am not clear on how to do so. Guide me how to do those in `benchmarks/benchmarks/linalg.py`?

**Comment by rgommers** [comment]
> So, to replicate those same benchmarking steps, I would then need to keep both the versions.

You don't have to. You only need to benchmark the ``circulant`` function. The way the benchmarking system works is that it can be run for any revision, and the result of each benchmark will be stored. Then when it's run for a future revision, the new result is stored. If ``time_new > time_old`` (with some tolerance), it is flagged as a performance regression. Results will be like so: http://pv.github.io/scipy-bench/

So concretely:
- after line 12, add ``from scipy.linalg import circulant`` (it's within the try-except to ensure that benchmarks run on old revisions as well, ``circulant`` was added in 0.16.0).
- add a new benchmark with ``def time_circulant``, followed by the code you want to time. The ``params`` thing you see in some other benchmarks is optional, that's for if you want to for example run the same benchmarks with different size input arrays.

**Comment by pv** [comment]
Benchmarks added.

**Comment by ilayn** [comment]
@droyed I think if we fix the reference issue this is good to go. Would you agree?

File: scipy/linalg/special_matrices.py [hunk 1]
```diff
@@ -236,13 +235,18 @@ def circulant(c):
...
-    a, b = np.ogrid[0:len(c), 0:-len(c):-1]
-    indx = a + b
-    # `indx` is a 2D array of indices into `c`, arranged so that `c[indx]` is
-    # the circulant matrix.
-    return c[indx]
+    # Form an extended array that could be strided to give circulant version
+    c_ext = np.concatenate((c[::-1], c[:0:-1]))
+    L = len(c)
+    n = c_ext.strides[0]
+    return as_strided(c_ext[L-1:], (L, L), (-n, n)).copy()
```
Comment (WarrenWeckesser) [inline scipy/linalg/special_matrices.py]: Use the explicit keyword arguments, like you did in `hankel` and `toeplitz`:

    return as_strided(c_ext[L-1:], shape=(L, L), strides=(-n, n)).copy()

**Comment by WarrenWeckesser** [comment]
I made one in-line comment, and I also agree that the reference to the stackoverflow question is not needed and could be removed.  Other than that, this looks good to me.

**Comment by droyed** [comment]
@pv Seems like good to go, except for the conflict at - `benchmarks/asv.conf.json`. Can you help out on resolving that?

File: benchmarks/asv.conf.json [hunk 2]
```diff
@@ -29,7 +29,7 @@
     "matrix": {
         "numpy": ["1.8.2"],
         "Tempita": ["0.5.2"],
-        "Cython": ["0.23.4"],
+        "Cython": ["0.25.2"],
```
Comment (tylerjereddy) [inline benchmarks/asv.conf.json]: @droyed The source of the merge conflict is the older Cython version used here. If you can update that string to `0.27.3` (we recently bumped the min version used), this should resole the issue (@pv could be quite busy & that should be a straightforward fix).

**Comment by tylerjereddy** [comment]
I went ahead and used the github GUI to resolve this straightforward `asv` configuration merge conflict.

**Comment by ilayn** [comment]
LGTM, in it goes. Thanks @droyed

**Comment by droyed** [comment]
@ilayn Awesome! Thanks.

**Pull request merged**